File: protein/old_Assign/Assign_2/scala/pRho.scala
//This will use strings instead of Vectors for rho
import scala.collection.immutable.Vector.empty
import scala.io.Source
import java.io.File
import math._
import util.Random
import Gamma.Gamma._

object pRho{
  val ko = 8; val d = 4;
  val k = ko/.62
  val theta = Array(.170207, .224021, .178348, .427424)
  def countEta (r:String ,s:Char):Int={
    var cnt = 0; if (r.charAt(0)==s) {cnt = 1}
    for (i <- 1 to (r.length-1)){
      if ( r.charAt(i-1) != r.charAt(i) ) {
        if (r.charAt(i)==s){ cnt+=1 }
      }
    } 
    cnt
  } 
  def inEta (rv:String):Array[Int]={
    val mv = Array(countEta(rv,'H'),countEta(rv,'E'),
                   countEta(rv,'T'),countEta(rv,'C')) 
    mv
  }
  def getEL (rv:String):Vector[(Char,Int)]={
    val rv2 = rv:+'A'
    val l = rv2.length; var i = 0
    var el = Vector.fill(0)('A',0)
    while (i < (l-1)){
      var j = 0
      while ( (rv2(i+j)==rv2(i+j+1)) & (i+j+2 < l) ){ j += 1 }
      el = el :+ (rv2(i),j+1)
      while ( (rv2.charAt(i+j)==rv2.charAt(i+j+1)) & (i+j+2 < l) ){ j += 1 }
      el = el :+ (rv2.charAt(i),j+1)
      i += j+1
    }
    el
  }
  def logNegBinom(x:Double, mu:Double, size:Double):Double={
    val p = size/(size+mu) 
    val n = size
    logGamma(x+n)-logGamma(n)-logGamma(x+1)+n*log(p)+x*log(1-p)
  }
  def sumLogLamGivenEta (EL:Vector[(Char,Int)]):Double={
    def logLamGivenEta (el: (Char,Int)):Double={
      if ((el._1=='H')&(el._2>=5)){logNegBinom(el._2-5,mu=1.885880,size=6.953392)}
      else if ((el._1=='E')&(el._2>=3)){logNegBinom(el._2-3,mu=2.521091,size=2.899121)}
      else if ((el._1=='T')&(el._2>=3)){logNegBinom(el._2-3,mu=0.839557,size=0.728294)}
      else if ((el._1=='C')&(el._2>=1)){logNegBinom(el._2-1,mu=0.990796,size=3.725501)}
      else {0}
    }
    var sumLog = 0.0
    for (i <- 0 to EL.size-1){
      sumLog = sumLog+logLamGivenEta(EL(i))
    }
    sumLog
  } 
  def logNorm(x:Double,mu:Double,sig:Double):Double={
    -.5*(log(2*Pi*sig*sig)+pow((x-mu)/sig,2))
  } 
  def logMult(mv:Array[Int],prob:Array[Double]):Double={
    var sum = 0.0
    for (i <- 0 to prob.size-1){
      sum = sum - logGamma(mv(i)+1) + mv(i)*log(prob(i))
    }
    sum+logGamma(1+mv.sum)
  }

  def main(args: Array[String]){
    val rho = "CEEETTTC" 
    val m = inEta(rho).sum
    val mv = inEta(rho).dropRight(1):+(inEta(rho)(3)-2)
    val EL = getEL(rho)
    val ans = exp(logNorm(m,2.347253+.154154*k, 5.526382/d)+logMult(mv,theta)+sumLogLamGivenEta(EL))
    println(ans)
  }

}
//To Compile: mkdir classes
//To Compile: scalac -d classes Gamma.scala
//To Compile:  scalac -d classes pRho.scala 
//To Execute:  scala -cp classes pRho 

File: protein/old_Assign/Assign_1/pRho.R
th <- c(.170207, .224021, .178348, .427424)

cntEta <- function(e, s){ 
  cnt = ifelse(e[1] == s, 1, 0)
  for (i in 2:length(e)){
    if ( (e[i-1] != e[i]) ){
      if (e[i]==s){cnt = cnt +1}
    }
  }
  return(cnt)
}

inEta <- function(r.v){
  m.v <- c(cntEta(r.v,'H'), cntEta(r.v,'E'), cntEta(r.v,'T'),cntEta(r.v,'C'))
  m <- sum(m.v) - 2
  m.v <- c(m.v[1:3],m.v[4]-2)
  return ( list( m.v, m ) )
}

getEL <- function(r.v){
  E <- NULL; L <- NULL
  l <-length(r.v); cnt = 0; i = 1
  while ( i <= l ){
    j = 0
    while( (r.v[i+j] == r.v[i+j+1]) & (i+j+1 <= l) ){
      j=j+1
    }
    cnt = cnt + 1
    E[cnt] <- r.v[i]; L[cnt] <- j+1
    i=i+j+1;
  }
  return ( list(E,L) )
}

LgivenE <- function(l,e){
  if (e=='H'){return (dnbinom(l-5,mu=1.885880,size=6.953392))}
  else if (e=='E'){return (dnbinom(l-3,mu=2.521091,size=2.899121))}
  else if (e=='T'){return (dnbinom(l-3,mu=0.839557,size=0.728294))}
  else {return (dnbinom(l-1,mu=0.990796,size=3.725501))}
}

LGE <- function(lam,eta){
  LgE <- NULL
  for (i in 1:length(lam)){
    LgE[i] <- LgivenE(lam[i],eta[i])
  }
  LgE
}

pRho <- function(rho,k=length(rho),d=4){
  m <- inEta(rho)[[2]] + 2
  mv <- inEta(rho)[[1]]
  E <- getEL(rho)[[1]]; L <- getEL(rho)[[2]]
  sumLogLamGivenEta <- 0; lge <- LGE(L,E)
  for (i in 1:length(E)){
    sumLogLamGivenEta <- sumLogLamGivenEta + log(lge[i])
  }
  
  exp(dnorm(m, 2.347253+.154154*k/.62, 5.526382/d,log=T) + dmultinom(mv,m-2,th,log=T) + sumLogLamGivenEta)
}
# for p=8, pRho is proportional to 1/.022483900175658
